bbc/simorgh
# Event tracking on top stories component

Resolves https://github.com/bbc/simorgh/issues/9082

**Overall change:**
Adds view and click event tracking mechanisms to the top stories component.

**Technical implementation:**
Creates a new React context called `EventTrackingContext` that is composed with other Simorgh contexts in the `withContexts` high order component which wraps all page components. `EventTrackingContext` provides all child components with most of the necessary props needed to be able to send an event to ATI. Custom hooks `useViewTracker`and `useClickTrackerHandler` receive the props from the context along with specific campaign and component type based props that needs passed into each hook call. When the view and click event mechanisms are triggered then all props are passed into `sendEventBeacon` which makes the call to ATI to log the event with all of the event data.

**Code changes:**

- updates `useViewTracker` hook API and [docs](https://github.com/bbc/simorgh/blob/click-view-hook/src/app/hooks/useViewTracker/README.md)
- updates `useClickTrackerHandler` hook API and [docs](https://github.com/bbc/simorgh/blob/click-view-hook/src/app/hooks/useClickTrackerHandler/README.md)

---

- [x] (BBC contributors only) This PR follows the [repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)

**Testing:**
- Go to any story page for any service and view the top stories component and ensure a request to ATI is sent with the correct tracking values e.g.:
<img width="720" alt="Screenshot 2021-05-21 at 09 14 30" src="https://user-images.githubusercontent.com/4798332/119116621-50aa7400-ba20-11eb-8b0e-d04fbb993133.png">

- Go to any story page for any service and click on a top stories link and ensure a request to ATI is sent with the correct tracking values.
- Ensure the view event tracking data and the click event tracking data is the same with the exception of the `ati` for the view event and `atc` for the click event.
- Check Safari bug - from this issue https://github.com/bbc/simorgh/issues/9070
> 🐛 There is an issue with ATI for Safari where navigating back after clicking on a tracked link triggers the link again - story BBCHOME-398. We must consider a workaround for this. WebCore have fixed this in their implementation - refer to their workaround for this.



- [x] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false yarn test:e2e:interactive`)
